rfcs/0031-ppc64-platforms.md

```diff
@@ -0,0 +1,86 @@
+This has already been merged in [nixpkgs#45340](https://github.com/NixOS/nixpkgs/pull/45340).
+
+The author of this RFC is already running a NixOS-based system on his POWER9-based computer, so for a basic system without graphical interface, only a handful of packages are known to require changes to build a working system:
+
+* mesa: currently assumes that any non-ARM system is x86. Requires restructuring the driver selection, but to the author's knowledge (since no actual graphical software was tested yet) it should not require any changes to the actual build description. ([nixpkgs#45474](https://github.com/NixOS/nixpkgs/pull/45474))
+* strace: also assumes that any non-ARM system is x86 and so tries to use the m32 personality. Requires checking the available machine personalities for all non-x86 architectures. ([nixpkgs#45472](https://github.com/NixOS/nixpkgs/pull/45472), already merged)
+* TeX Live: builds LuaJIT as part of mfluajit, which doesn't work because LuaJIT doesn't support 64-bit Power architectures yet; seems to be possible to disable on that platform without too much impact. ([nixpkgs#45475](https://github.com/NixOS/nixpkgs/pull/45475))
+* Bootloader tooling:
+    * the internal option *system.boot.loader.kernelFile* is currently set to the kernelTarget attribute of the host platform definition.
+      PowerPC kernels however use `zImage` as build target, yet produce a `vmlinux` file.
+      A one-line change would allow the option to default to the host platform definition's kernelFile attribute, if set.
+    * A typical bootloader for OpenPOWER systems is Petitboot. It can read a variety of bootloader configuration files, including those of GRUB 2 and SYSLINUX.
+      Using GRUB, even with *boot.loader.grub.device* = "nodev", as it is, implies building GRUB for the host architecture, which fails for ppc64le (currently because it requires soft-float).
+
+      However, the generic-extlinux-compatible bootloader module can be modified to provide options to store the configuration files in a path that Petitboot will consider and use absolute paths starting from the filesystem root, so that Petitboot can load the kernel/initrd images.
+
+Beyond a headless system, the biggest obstacle for ppc64(le) support seems to be Rust, because a lot of desktop software has it as a (direct or transitive) build dependency.
+In particular, a lot of GNOME software has a dependency on librsvg (which has parts written in Rust) via their documentation tools.
+The author of this RFC plans to create a way to build bootstrapping binaries for Rust on ppc64le (either via cross-building from x86 or from source via mrustc).
+
+Some packages, like OpenBLAS, will need target descriptions for these platforms, but many don't require changes to the actual build process.
+
+Support for Big Endian ppc64 might require changes in a few packages, though the majority of changes are assumed to be upstreamable (see [Unresolved Questions][unresolved]).
```

Comment:
How about keeping track of this progress in the nixos.wiki ?

While these information give an idea of how much work is left to do, the RFC should also cover other aspects of adding new platforms like the added cost of nix evaluation and maintenance overhead that it creates inside of nixpkgs. I would like to see things addressed like:

* what are the rules of package maintainers in interaction with the non-official platforms?
* where is the list of official and unofficial platforms listed?
* what are the rules for adding and dropping a platform?
* for non-official platforms, are they keeping their own build farm? if yes, how can this be listed?
* where should the stdenv bootstrap files be stored?

Reply:
One thing that I would like to see explored, is if it's possible to store the stdenv out of the main nixpkgs tree, along with the binary cache and all the platform-specific information. Using the @nix-community organisation would be OK for that type of activity if you want.

Reply:
> One thing that I would like to see explored, is if it's possible to store the stdenv out of the main nixpkgs tree

Would this make Continuous Integration less easy?

One thing I'd definitely want to have as a maintainer is swift feedback on whether changes I make break things for some architecture.

Reply:
how would you be notified that your change broke an architecture that is not checked by ofborg and the main hydra?

let's say I am the maintainer of POWER, I can:
* add power to the main nixpkgs architectures
* create my own nixpkgs-power repo. that repo contains the stdenv, overrides and power-specific packages
* create my own hydra instance that runs whenever nixpkgs master or nixpkgs-power changes => and publishes my binary cache